wasm/src/lsp/types.rs

```diff
@@ -0,0 +1,300 @@
+use std::cmp::PartialOrd;
+use std::path::PathBuf;
+use std::str::FromStr;
+use std::rc::Rc;
+
+use lsp_server::{ExtractError, Request, RequestId};
+
+use lsp_types::{
+    Position,
+    Range, SemanticTokenModifier, SemanticTokenType
+};
+
+use percent_encoding::percent_decode;
+use url::{Host, Url};
+
+use clvm_tools_rs::compiler::sexp::decode_string;
+use clvm_tools_rs::compiler::srcloc::Srcloc;
+
+lazy_static! {
+    pub static ref TOKEN_TYPES: Vec<SemanticTokenType> = {
+        vec![
+            SemanticTokenType::PARAMETER,
+            SemanticTokenType::VARIABLE,
+            SemanticTokenType::FUNCTION,
+            SemanticTokenType::MACRO,
+            SemanticTokenType::KEYWORD,
+            SemanticTokenType::COMMENT,
+            SemanticTokenType::STRING,
+            SemanticTokenType::NUMBER,
+        ]
+    };
+    pub static ref TOKEN_MODIFIERS: Vec<SemanticTokenModifier> = {
+        vec![
+            SemanticTokenModifier::DEFINITION,
+            SemanticTokenModifier::READONLY,
+            SemanticTokenModifier::DOCUMENTATION,
+        ]
+    };
+}
+
+pub const TK_PARAMETER_IDX: u32 = 0;
+pub const TK_VARIABLE_IDX: u32 = 1;
+pub const TK_FUNCTION_IDX: u32 = 2;
+pub const TK_MACRO_IDX: u32 = 3;
+pub const TK_KEYWORD_IDX: u32 = 4;
+pub const TK_COMMENT_IDX: u32 = 5;
+pub const TK_STRING_IDX: u32 = 6;
+pub const TK_NUMBER_IDX: u32 = 7;
+
+pub const TK_DEFINITION_BIT: u32 = 0;
+pub const TK_READONLY_BIT: u32 = 1;
+
+pub struct ToFilePathErr;
+
+// Note: to_file_path is only present on native builds, but we're building to
+// wasm.
+//
+// The code isn't ungeneric, so we'll express it here.
+pub trait HasFilePath {
+    fn our_to_file_path(&self) -> Result<PathBuf, ToFilePathErr>;
+}
+
+fn file_url_segments_to_pathbuf(
```

Comment:
I see - we can't use a library because we are emitting wasm?

Is it possible for a zero-length `segments` to come in here?

Reply:
e.g. `file:///`

Reply:
good point, will add a test for that.